src/lightning/data/utilities/env.py

```diff
@@ -37,7 +38,10 @@ def detect(cls) -> "_DistributedEnv":
         if world_size is None or world_size == -1:
             world_size = 1
 
-        return cls(world_size=world_size, global_rank=global_rank)
+        # TODO: Add support for other accelerators
+        num_nodes = (world_size // torch.cuda.device_count()) if torch.cuda.is_available() else 1
+
```

Comment:
This is not very robust. This can lead to silent misbehavior. At the very least we should error and not just round down to the nearest integer. 

But moreover, let's read the num nodes from the environment variable, not the torch device count?

Reply:
I added a divisible assertion to avoid errors. I think this is enough now as it is well defined on the platform.